File: lib/features/live/logic/quick_message_policy.dart
/// Summary:
/// - Client-side validation for Premium custom quick messages.
/// - Best-effort anti-abuse (cannot guarantee against modified clients).
class QuickMessagePolicy {
  // Your rules (from user):
  static const int maxCustomCount = 15;
  static const int maxChars = 15;
  static const bool allowEmojis = true;
  static const bool allowUrlsAndMentions = false;

  // Safe defaults fallback (English).
  static const List<String> defaultFallback = <String>[
    'Focus!',
    'Calm down',
    'We got this',
    'One more goal!',
    'Don’t give up',
    'Sorry',
    'Unlucky',
    'What a goal!',
    'Ref??',
  ];

  // Best-effort profanity/abuse blocklist (English-centric).
  // Keep this list minimal to reduce false-positives; expand as needed.
  static const List<String> _blockedTokens = <String>[
    'fuck',
    'shit',
    'bitch',
    'asshole',
    'bastard',
    'dick',
    'pussy',
    'cunt',
  ];

  static final RegExp _urlLike = RegExp(r'(https?:\/\/|www\.|\.com\b|\.net\b|\.org\b)', caseSensitive: false);
  static final RegExp _mentionLike = RegExp(r'@[\w_]+', caseSensitive: false);

  // Loose "phone-number like" pattern (optional).
  static final RegExp _phoneLike = RegExp(r'(\+?\d[\d\-\s]{6,}\d)');

  static String normalize(String raw) {
    // Trim and collapse repeated whitespace.
    final s = raw.trim().replaceAll(RegExp(r'\s+'), ' ');
    return s;
  }

  static QuickMessageValidationResult validateCustomMessage(String raw) {
    final s = normalize(raw);
    if (s.isEmpty) return const QuickMessageValidationResult.invalid('Message is empty');
    if (s.runes.length > maxChars) return QuickMessageValidationResult.invalid('Max $maxChars characters');

    if (!allowUrlsAndMentions) {
      if (_urlLike.hasMatch(s)) return const QuickMessageValidationResult.invalid('Links are not allowed');
      if (_mentionLike.hasMatch(s)) return const QuickMessageValidationResult.invalid('Mentions are not allowed');
      if (_phoneLike.hasMatch(s)) return const QuickMessageValidationResult.invalid('Phone numbers are not allowed');
    }

    final lower = s.toLowerCase();
    for (final t in _blockedTokens) {
      if (lower.contains(t)) {
        return const QuickMessageValidationResult.invalid('Message is not allowed');
      }
    }

    // If emojis not allowed, restrict to basic punctuation/letters/numbers/spaces.
    if (!allowEmojis) {
      final ok = RegExp(r"^[a-zA-Z0-9\s\.\,\!\?\-\’\'\"]+$").hasMatch(s);
      if (!ok) return const QuickMessageValidationResult.invalid('Only plain text is allowed');
    }

    return QuickMessageValidationResult.valid(s);
  }

  static List<String> sanitizeList(List<String> raw) {
    // Apply normalize + remove empty + trim to max count.
    final cleaned = <String>[];
    for (final m in raw) {
      final v = normalize(m);
      if (v.isEmpty) continue;
      cleaned.add(v);
      if (cleaned.length >= maxCustomCount) break;
    }
    return cleaned;
  }
}

class QuickMessageValidationResult {
  final bool ok;
  final String value;
  final String error;

  const QuickMessageValidationResult._(this.ok, this.value, this.error);

  const QuickMessageValidationResult.valid(String value) : this._(true, value, '');
  const QuickMessageValidationResult.invalid(String error) : this._(false, '', error);
}

File: web_client/lib/services/quickMessagePolicy.ts
// Mirrors lib/features/live/logic/quick_message_policy.dart's
// QuickMessagePolicy exactly — same limits, same blocked patterns, same
// default fallback list — so a message that validates on one platform
// validates on the other, and the shared Firestore write
// (users/{uid}.quickMessagesCustom) never gets a value one platform
// would have rejected.

export const QUICK_MESSAGE_MAX_COUNT = 15;
export const QUICK_MESSAGE_MAX_CHARS = 15;

export const QUICK_MESSAGE_DEFAULT_FALLBACK: string[] = [
  'Focus!',
  'Calm down',
  'We got this',
  'One more goal!',
  'Don’t give up',
  'Sorry',
  'Unlucky',
  'What a goal!',
  'Ref??',
];

const BLOCKED_TOKENS = ['fuck', 'shit', 'bitch', 'asshole', 'bastard', 'dick', 'pussy', 'cunt'];

const URL_LIKE = /(https?:\/\/|www\.|\.com\b|\.net\b|\.org\b)/i;
const MENTION_LIKE = /@[\w_]+/i;
const PHONE_LIKE = /(\+?\d[\d\-\s]{6,}\d)/;

export interface QuickMessageValidationResult {
  ok: boolean;
  value: string;
  error: string;
}

export function normalizeQuickMessage(raw: string): string {
  return raw.trim().replace(/\s+/g, ' ');
}

export function validateCustomQuickMessage(raw: string): QuickMessageValidationResult {
  const s = normalizeQuickMessage(raw);
  if (!s) return { ok: false, value: '', error: 'Message is empty' };
  if ([...s].length > QUICK_MESSAGE_MAX_CHARS) {
    return { ok: false, value: '', error: `Max ${QUICK_MESSAGE_MAX_CHARS} characters` };
  }
  if (URL_LIKE.test(s)) return { ok: false, value: '', error: 'Links are not allowed' };
  if (MENTION_LIKE.test(s)) return { ok: false, value: '', error: 'Mentions are not allowed' };
  if (PHONE_LIKE.test(s)) return { ok: false, value: '', error: 'Phone numbers are not allowed' };

  const lower = s.toLowerCase();
  if (BLOCKED_TOKENS.some((t) => lower.includes(t))) {
    return { ok: false, value: '', error: 'Message is not allowed' };
  }

  return { ok: true, value: s, error: '' };
}

export function sanitizeQuickMessageList(raw: string[]): string[] {
  const cleaned: string[] = [];
  for (const m of raw) {
    const v = normalizeQuickMessage(m);
    if (!v) continue;
    cleaned.push(v);
    if (cleaned.length >= QUICK_MESSAGE_MAX_COUNT) break;
  }
  return cleaned;
}
